File: src/main/kotlin/com/elissamc/core/PlayerProfile.java
package com.elissamc.core;

import cn.nukkit.Player;
import cn.nukkit.Server;
import cn.nukkit.utils.Config;
import com.google.common.collect.ImmutableList;

import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlID;
import java.io.File;
import java.util.*;

public class PlayerProfile {
    /**
     * Player's UUID.
     */
    @XmlID
    @XmlAttribute
    protected final UUID uuid;
    /**
     * Player's friends.
     */
    @XmlAttribute
    protected final List<UUID> friends = new ArrayList<UUID>();
    /**
     * Player's foes.
     */
    @XmlAttribute
    protected final List<UUID> foes = new ArrayList<UUID>();
    /**
     * Player's settings.
     */
    @XmlAttribute
    protected final Map<Settings, Boolean> settings = new HashMap<Settings, Boolean>();

    /**
     * Spectating status.
     */
    protected transient boolean spectating = false;

    /**
     * Last known name of this player.
     */
    @XmlAttribute
    protected String lastKnownName = "";
    /**
     * Amount of player's points (probably in-game currency or whatever).
     */
    @XmlAttribute
    protected int points = 0;
    @XmlAttribute
    protected int warnCount = 0;

    /**
     * Represents party, that player is currently in. If is player not in any party, it is null.
     */
    protected Party party = new Party(null);

    /**
     * Creates player profile from UUID.
     *
     * @param player
     */
    PlayerProfile(final UUID player) {
        this.uuid = player;
        for(Player p : Server.getInstance().getOnlinePlayers().values()){
            if(p.getUniqueId() == player)
                this.lastKnownName = p.getName();
        }
    }

    /**
     * Adds friend.
     *
     * @param player
     */
    public void addFriend(final UUID player) {
        this.friends.add(player);
    }

    /**
     * Removes friend.
     *
     * @param player
     */
    public void removeFriend(final UUID player) {
        this.friends.remove(player);
    }

    public void addFoe(final UUID player) {
        this.foes.add(player);
    }

    public void removeFoe(final UUID player) {
        this.foes.remove(player);
    }

    /**
     * Returns UUID of profile.
     *
     * @return
     */
    UUID getUniqueId() {
        return this.uuid;
    }

    /**
     * Return list of player's friends.
     *
     * @return
     */
    List<UUID> getFriends() {
        return ImmutableList.copyOf(this.friends);
    }

    List<UUID> getFoes() {
        return ImmutableList.copyOf(this.foes);
    }

    /**
     * Returns whatever is this player friend with the specified.
     *
     * @param uniqueId Friend's UUID.
     * @return Return boolean of whether friends or not.
     */
    public boolean isFriend(final UUID uniqueId) {
        return this.friends.contains(uniqueId);
    }

    public boolean isFoe(final UUID uniqueId) {
        return this.foes.contains(uniqueId);
    }

    /**
     * Returns setting value.
     *
     * @param setting Setting string.
     * @return Returns boolean.
     */
    boolean getSetting(final Settings setting) {
        return this.settings.getOrDefault(setting, true);
    }

    /**
     * Set players settings.
     *
     * @param setting Setting string name.
     * @param value True or False value.
     */
    public void setSetting(final Settings setting, final boolean value) {
        this.settings.put(setting, value);
    }

    /**
     * Return's whether is player spectating.
     *
     * @return
     */
    public boolean isSpectating() {
        return this.spectating;
    }

    /**
     * Sets player's spectating state.
     *
     * @param spectating Set player as spectating.
     */
    public void setSpectating(final boolean spectating) {
        this.spectating = spectating;
    }

    /**
     * Saves player's profile to file.
     *
     * @param path path to save.
     */
    void save(final String path) {
        Config yaml = new Config(path, Config.YAML);

        yaml.set("player.uuid", this.uuid.toString());
        yaml.set("player.points", this.points);
        yaml.set("player.warnCount", this.warnCount);
        yaml.set("player.lastKnownName", this.lastKnownName);
        yaml.set("player.friends", this.friends);
        yaml.set("player.foes", this.foes);

        yaml.save(new File(path));
    }

    /**
     * Loads player profile from specified path.
     *
     * @param path Path to player profile.
     */
    static PlayerProfile load(final String path) {
        Config yaml = new Config(path, Config.YAML);
        UUID uuid = UUID.fromString(yaml.getString("player.uuid"));

        PlayerProfile profile = new PlayerProfile(uuid);

        profile.points = yaml.getInt("player.points");
        profile.lastKnownName = yaml.getString("player.lastKnownName");

        List<?> friends = yaml.getList("player.friends");
        List<?> foes = yaml.getList("player.foes");

        for (Object obj : friends)
            profile.addFriend(UUID.fromString(obj.toString()));
        for (Object obj : foes)
            profile.addFoe(UUID.fromString(obj.toString()));

        return profile;
    }

    /**
     * Return the number of points.
     *
     * @return Returns player's # of points.
     */
    public int getPoints() {
        return this.points;
    }

    /**
     * Add specified amount of points.
     *
     * @param points Add points to player.
     */
    public void addPoints(final int points) {
        this.points += points;
    }

    public Party getParty() {
        return this.party;
    }

    public void setParty(final Party party) {
        this.party = party;
    }
}

File: src/main/kotlin/com/elissamc/core/PlayerProfile.kt
package com.elissamc.core

import cn.nukkit.Server
import cn.nukkit.utils.Config
import com.google.common.collect.ImmutableList
import java.io.File
import java.util.*
import javax.xml.bind.annotation.XmlAttribute
import javax.xml.bind.annotation.XmlID

open class PlayerProfile
/**
 * Creates player profile from UUID.
 *
 * @param uniqueId
 */
internal constructor(
        /** Player's UUID. */
        @field:XmlID
        @field:XmlAttribute
        internal val uniqueId: UUID) {

    /** Player's friends. */
    @XmlAttribute
    private val friends: MutableList<UUID> = ArrayList()

    /** Player's foes. */
    @XmlAttribute
    private val foes: MutableList<UUID> = ArrayList()

    /** Player's settings. */
    @XmlAttribute
    private val settings: MutableMap<Settings, Boolean> = HashMap()

    /** Spectating status. */
    @Transient
    var isSpectating = false

    /** Last known name of this player. */
    @XmlAttribute
    protected var lastKnownName = ""

    /** Amount of player's points (probably in-game currency or whatever). */
    @XmlAttribute
    var points = 0
        protected set

    /** Amount of warning's the player has */
    @XmlAttribute
    private var warnCount = 0

    /** Represents party, that player is currently in. If is player not in any party, it is null. */
    var party: Party? = null

    init {
        this.lastKnownName = uniqueId.toString()
        for (p in Server.getInstance().onlinePlayers.values) {
            if (p.uniqueId === uniqueId)
                this.lastKnownName = p.name
        }
    }

    /**
     * Adds friend.
     *
     * @param player
     */
    fun addFriend(player: UUID) {
        this.friends.add(player)
    }

    /**
     * Removes friend.
     *
     * @param player
     */
    fun removeFriend(player: UUID) {
        this.friends.remove(player)
    }

    fun addFoe(player: UUID) {
        this.foes.add(player)
    }

    fun removeFoe(player: UUID) {
        this.foes.remove(player)
    }

    /**
     * Return list of player's friends.
     *
     * @return
     */
    internal fun getFriends(): List<UUID> {
        return ImmutableList.copyOf(this.friends)
    }

    internal fun getFoes(): List<UUID> {
        return ImmutableList.copyOf(this.foes)
    }

    /**
     * Returns whatever is this player friend with the specified.
     *
     * @param uniqueId Friend's UUID.
     * @return Return boolean of whether friends or not.
     */
    fun isFriend(uniqueId: UUID): Boolean {
        return this.friends.contains(uniqueId)
    }

    fun isFoe(uniqueId: UUID): Boolean {
        return this.foes.contains(uniqueId)
    }

    /**
     * Returns setting value.
     *
     * @param setting Setting string.
     * @return Returns boolean.
     */
    internal fun getSetting(setting: Settings): Boolean {
        return (this.settings as Map<Settings, Boolean>).getOrDefault(setting, true)
    }

    /**
     * Set players settings.
     *
     * @param setting Setting string name.
     * @param value True or False value.
     */
    fun setSetting(setting: Settings, value: Boolean) {
        this.settings[setting] = value
    }

    /**
     * Saves player's profile to file.
     *
     * @param path path to save.
     */
    internal fun save(path: String) {
        val yaml = Config(path, Config.YAML)

        val friends = ArrayList<String>()
        this.friends.forEach { uuid -> friends.add(uuid.toString()) }
        val foes = ArrayList<String>()
        this.foes.forEach { uuid -> foes.add(uuid.toString()) }

        yaml.set("player.uuid", this.uniqueId.toString())
        yaml.set("player.points", this.points)
        yaml.set("player.warnCount", this.warnCount)
        yaml.set("player.lastKnownName", this.lastKnownName)
        yaml.set("player.friends", friends)
        yaml.set("player.foes", foes)

        yaml.save(File(path))
    }

    /**
     * Add specified amount of points.
     *
     * @param points Add points to player.
     */
    fun addPoints(points: Int) {
        this.points += points
    }

    companion object {

        /**
         * Loads player profile from specified path.
         *
         * @param path Path to player profile.
         */
        internal fun load(path: String): PlayerProfile {
            val yaml = Config(path, Config.YAML)
            val uuid = UUID.fromString(yaml.getString("player.uuid"))

            val profile = PlayerProfile(uuid)

            profile.points = yaml.getInt("player.points")
            profile.lastKnownName = yaml.getString("player.lastKnownName")

            val friends = yaml.getStringList("player.friends")
            val foes = yaml.getStringList("player.foes")

            for (obj in friends)
                profile.addFriend(UUID.fromString(obj.toString()))
            for (obj in foes)
                profile.addFoe(UUID.fromString(obj.toString()))

            return profile
        }
    }
}